src/frontier_detection/src/map.cpp

```diff
@@ -0,0 +1,278 @@
+                    RCLCPP_WARN(
+                        logger,
+                        "Invalid node detected. Coordinates: [%f, %f, %f], Occupancy: NaN",
+                        it.getCoordinate().x(),
+                        it.getCoordinate().y(),
+                        it.getCoordinate().z()
+                    );
+
+                    // Remove the invalid node or reset its value
+                    globalMap->deleteNode(it.getKey());  // Remove the node
+                }
+            }
+            return true;
+        }));
+        index += chunk_size;
+    }
+
+    // Wait for all tasks and aggregate results
+    for (auto& task : tasks) {
+        if (!task.get()) {
+            return false; // If any thread fails validation
+        }
+    }
+
+    return true; // All nodes are valid
+}
+
+OctoMapNode::OctoMapNode(std::string nodeName): Node(nodeName){
+	RCLCPP_INFO(this->get_logger(), "Creating OctoMap Node");
+	mpGlobalMap = std::make_shared<OctoMap>(this->get_logger());
+    const int num_channels = 3; // x y z
+    mpRawPointCloud.header.frame_id = "map";
+    mpRawPointCloud.height = 1;
+    mpRawPointCloud.width = 0;
+    mpRawPointCloud.is_bigendian = false;
+    mpRawPointCloud.is_dense = true;
+    mpRawPointCloud.point_step = num_channels * sizeof(float);
+    mpRawPointCloud.row_step = mpRawPointCloud.point_step * mpRawPointCloud.width;
+    mpRawPointCloud.fields.resize(num_channels);
+    std::string channel_id[] = { "x", "y", "z"};
+    for (int i = 0; i<num_channels; i++) {
+        mpRawPointCloud.fields[i].name = channel_id[i];
+        mpRawPointCloud.fields[i].offset = i * sizeof(float);
+        mpRawPointCloud.fields[i].count = 1;
+        mpRawPointCloud.fields[i].datatype = sensor_msgs::msg::PointField::FLOAT32;
+    }
+
+    mpRawPointCloud.data.resize(mpRawPointCloud.row_step * mpRawPointCloud.height);
+	mpPointCloudSubscriber = this->create_subscription<MapMsg>(
+			"/orbslam3_mono_node/map_points",
+			10,
+			std::bind(&OctoMapNode::AddNewPointCloudCallback, this, std::placeholders::_1));
+	mpMapPublisher = this->create_publisher<octomap_msgs::msg::Octomap>("~/octomap", 10);
+	mpPointCloudPublisher = this->create_publisher<sensor_msgs::msg::PointCloud2>("~/point_cloud", 10);
+	mpMapPublisherTimer = this->create_wall_timer(
+		std::chrono::milliseconds(100), std::bind(&OctoMapNode::PublishMap, this)
+	);
+	mpThrMsgQueueProcess = new std::thread(&OctoMapNode::UpdateMap, this);	
+}
+
+void OctoMapNode::UpdateMap(){
+	MapMsg::SharedPtr msg;
+	while(true){
+		{
+			std::unique_lock<std::mutex> lock(mpMtxMsgQueue);
+			if(mpMsgQueue.empty()){
+				continue;
+			}
+			else{
+				msg = mpMsgQueue.front();	
+				mpMsgQueue.pop_front();
+			}
+	
+		}
+		octomap::Pointcloud cloud;
+		octomap::point3d point;
+		octomap::pointCloud2ToOctomap(msg->cloud, cloud);
+		// ConvertPointCloud2ToOctomap(msg->cloud, cloud);
+		CreatePoint3dFromTransform(msg->pose, point);
+		mpGlobalMap->UpdateMap(cloud, point);
+	}
```

Comment:
can we sleep a bit, if not this thread will suck up a lot of compute

Reply:
and its also possible it will keep owning the lock before others get a chance

Reply:
UpdateMap may be a few ms long and since the mutex is unlocked due to end of the scope in which it was locked, I thought sleep is not needed. Let me check how long UpdateMap generally take and then decide on a sleep duration here.

Reply:
fair i had missed the scope brackets